File: leaderboard/personal/go/stats.go
package main

import "time"

type DurationType int
type IntType int

const (
	PreciseDuration  DurationType = iota
	OverADayDuration DurationType = iota
	MissingDuration  DurationType = iota
)

const (
	ValidInt   IntType = iota
	MissingInt IntType = iota
)

type IntStat struct {
	Type  IntType
	Value int
}

type DurationStat struct {
	Type     DurationType
	Duration time.Duration
}

type DayPartStats struct {
	Year  IntStat
	Day   IntStat
	Time  DurationStat
	Rank  IntStat
	Score IntStat
}

type DayStats struct {
	Year  IntStat
	Day   IntStat
	Parts [2]DayPartStats
}

File: leaderboard/personal/rs/src/stats.rs
use chrono;

#[derive(Debug)]
pub enum Integer {
    Some(usize),
    Missing,
    Unknown,
}

#[derive(Debug)]
pub enum Duration {
    Some(chrono::Duration),
    OverOneDay,
    Missing,
    Unknown,
}

#[derive(Debug, Clone)]
pub struct Date {
    pub day: usize,
    pub year: usize,
}

#[derive(Debug)]
pub struct DayPartStats {
    pub date: Date,
    pub part: usize,
    pub time: Duration,
    pub rank: Integer,
    pub score: Integer,
}

#[derive(Debug)]
pub struct DayStats {
    pub date: Date,
    pub parts: [DayPartStats; 2],
}
